### PR TITLE
TIMX 385 - browsable, filterable Records table

### PR DESCRIPTION
**Pull request by ghukill**
### Purpose and background context

This PR reworks how to review individual records from a run.  This is accomplished by removing the "Record Samples" section in the Run page, which pointed to a standalone, static HTML page of Record links, and replacing it with a **single, filterable, browsable table of all records**.

The previous approach was functional, but had some drawbacks.  The "Samples" page was static HTML, and did not lend itself to large numbers of records.  So, a "sample" of Records that matched the criteria was displayed, but this necessarily was not showing _all_ records that matched that critiera.  And, this approach had no inroads for viewing records with _no_ diff, where it could still be helpful to look at the A/B versions, even if identical. The list goes on.

The approach taken here is a single table, capable of handling huge numbers of records.  By layering on filtering and searching, it avoids the issue of thinking of combinations in advance of records to view.  It also -- by default -- shows all records, allowing for a more browsing experience.  The following is a screenshot of what it looks like:

Screenshot 2024-11-07 at 3 47 24 PM

The Javascript library [DataTables](https://datatables.net/) is used for this table, which is highly configurable.  The approach of "server-side" data preparation is used, so our Flask webapp is doing the heavy lifting of performing queries, sorting results, etc.  The actual table then just makes lightweight requests to the flask data endpoint each time information is needed, or filters applied, sorting, etc.  In this way, while the records dataset may contain 4-5m records, we'd only ever be sending 10-100 records (depending on "Records per Page" setting) to the table itself.

### How can a reviewer manually see the effects of these changes?

1- Set production AWS credentials

2- Init job (where `4fcb617` is an fairly old commit that generates some diffs):
```shell
pipenv run abdiff --verbose init-job -d output/jobs/records -a 4fcb617 -b 3f77c7c
```

3- Generate a CSV of input files and perform a run:
```shell
pipenv run abdiff --verbose timdex-sources-csv -o output/jobs/records/springshare.csv -s libguides,researchdatabases
pipenv run abdiff --verbose run-diff -d output/jobs/records -i output/jobs/records/springshare.csv
```

4- View the job:
```shell
pipenv run abdiff --verbose view-job -d output/jobs/records
```

Here are some things to try with the "Records" table at the bottom of the run page.  Recommended to click "Reset Filters" button between them (_NOTE: the file counts are approximate, given changes in records depending on when run_):
- Without any filters, note records count is 1,290
  - click source `libguides`, note count goes down to 372
  - hold ctrl (or cmd?) and click `researchdatabases`, note count goes _back up_ to 1,290 because effectively all records again
  - this is multi-entry selection, which works for fields as well
- Click source `researchdatabases` + field `publishers`, note that zero records match this combo
- Enter "Chemical Engineering" into the search box, note that two records show up
  - this was fulltext searching the actual A and B records, so not obvious where it matched, but clicking into either record will reveal it's somewhere in the record body
- Enter "libguides:guides-176330" in search box, one result comes back, showing how it could be used for `timdex_record_id` as well, as it's part of the source record 

### Includes new or updated dependencies?
YES: but only in the webapp, the DataTables javascript and CSS files

### Changes expectations for external applications?
NO

### What are the relevant tickets?
- https://mitlibraries.atlassian.net/browse/TIMX-385

### Developer
- [X] All new ENV is documented in README
- [X] All new ENV has been added to staging and production environments
- [X] All related Jira tickets are linked in commit message(s)
- [X] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer(s)
- [ ] The commit message is clear and follows our guidelines (not just this PR message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The provided documentation is sufficient for understanding any new functionality introduced
- [ ] Any manual tests have been performed **or** provided examples verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

File: abdiff/webapp/app.py [hunk 1]
```diff
@@ -89,63 +89,49 @@ def run(run_timestamp: str) -> str:
         except FileNotFoundError:
             transform_logs = "'logs.txt' not found for transform logs"
```
Comment (jonavellecuerdo) [inline abdiff/webapp/app.py:91]: Something that caught my attention when reviewing the web app is this: 
image

Will this only show the logs if `PRESERVE_ARTIFACTS="true"`? Can we move `logs.txt` to the `run/YYYY-MM-DD_HH-MM-SS/` directory, such that `logs.txt` is preserved even after `transformed/` directory is removed? 

The[ example-job-3 test fixture](https://github.com/MITLibraries/transmogrifier-ab-diff/tree/TIMX-385-records-preview/tests/fixtures/jobs/example-job-3/runs/2024-11-07_03-56-22) seems to suggest that this is what _should_ be happening. 🤔

**Review by jonavellecuerdo — COMMENTED** [review]
For my review, I followed the instructions outlined in the PR description, and I cannot express how awesome these changes are! I was amazed at how quickly it was able to run queries against the record dataset and how user-friendly the interface is. Excellent job. I did pose one question + potential request for change!

File: abdiff/webapp/app.py (hunk 1 above)
Comment (ghukill) [reply abdiff/webapp/app.py:91]: 100% agreed.  I _almost_ put this into the PR, but wanted to keep it tightly scoped to the records table.

What I should have done is acknowledge the issue, but pointed to these two open tickets:
- https://mitlibraries.atlassian.net/browse/TIMX-368
- https://mitlibraries.atlassian.net/browse/TIMX-382

As with many things in this project, would hate to apply a band-aid to what feels like needs a little updating and rethinking (given scale).  Are you okay @jonavellecuerdo just keeping as-is for the moment, and relying on these tickets to resolve?

**Comment by ghukill** [comment]
> For my review, I followed the instructions outlined in the PR description, and I cannot express how awesome these changes are! I was amazed at how quickly it was able to run queries against the record dataset and how user-friendly the interface is. Excellent job. I did pose one question + potential request for change!

Thanks!  FWIW, once we get into the 3-4m records, I think we'll see a _significant_ delay in the responsiveness of the table.  At that time, we can see if we need optimizations, or if it's just 1-2 seconds, perhaps that is okay for this diagnostic tool.

**Review by jonavellecuerdo — APPROVED** [review]

File: abdiff/webapp/app.py (hunk 1 above)
Comment (ehanson8) [reply abdiff/webapp/app.py:91]: Agree with that approach as well

**Review by ehanson8 — APPROVED** [review]
Looks great!

**Pull request merged**